### PR TITLE
Collapse reducers and effects into actions. Implement #126.

### PR DESCRIPTION
**Pull request by jbucaran**
**Remove** reducers/effects for a familiar concept: _actions_.

* Actions can be used to update the model/render the view. How? Return the new model, just like with reducers before:

```jsx
app({
    model: 0,
    actions: {
        add: model => model + 1,
        sub: model => model - 1
    }
})
```

* In addition, actions can be used to cause side effects or update the view asynchronously. How? You have two options:

1. Same as before. Run your async tasks inside the function and when you are done call further actions to update the view if necessary.
```jsx
app({
  actions: {
    add: model => model + 1,
    sub: model => model - 1,
    waitThenAdd: (model, data, actions) => {
      setTimeout(actions.add, 1000)
  }
})
```
2. Return a Promise. In this fashion, you can have multiple actions that return promises and chain them together using the [.then](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) or catch errors via [.catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch). 
```jsx
app({
  actions: {
    add: model => model + 1,
    sub: model => model - 1,
    waitThenAdd: (model, data, actions) =>
      new Promise(resolve => setTimeout(resolve, 1000)).then(actions.add)
  }
})
```
If you don't know Promises work or don't want to use them, that's okay. Promises **are not necessary** in order to use hyperapp, but it's nice you could use them if you wanted and hyperapp does the right thing (which is returning them to the caller instead of updating the model/view).

### Caveats

* Before, reducers and effects had different function signatures. Now all actions have the same signature: (model, data, actions, error).
* To know whether an action updates the model/view or not, Hyperapp checks its return value. If you return a primitive/model then it's logical to assume you want to update the model.
* If an action returns undefined, it doesn't make much sense to update the model or render the view.

**Related**:
* https://github.com/hyperapp/hyperapp/issues/126

### CODE REVIEW

**Comment by jbucaran** [comment]
@dodekeract Review please 🙏.

**Comment by lukejacksonn** [comment]
Have you seen my proposal [here](https://github.com/hyperapp/hyperapp/issues/126#issuecomment-284192686). I have only just skimmed this PR but I think we are trying to achieve the same thing.

File: src/app.js [hunk 1]
```diff
@@ -84,27 +70,35 @@ export default function (app) {
...
+    root = app.root || document.body.appendChild(document.createElement("div"))
+
+    render(model, view)
+
+    for (var i = 0; i < subscriptions.length; i++) {
+      subscriptions[i](model, actions, onError)
+    }
+  })
+
   function load(fn) {
```
Comment (jbucaran) [inline src/app.js]: @liadbiz Thanks! :) 

See #69 thread for the original reason we introduced this code.

**Comment by jbucaran** [comment]
I'd like to proceed with this PR 🔥.

Please if you have any further comments, post them soon 🙏 🙏.

**Comment by lukejacksonn** [comment]
I _think_ I am up to date with updates haha!

> Correct me if I am wrong, but your proposal requires actions to always return a Promise?

I'm assuming you are saying this because of `Promise.resolve(action(m,d,a))` if so then no this is not true as `Promise.resolve(42) /=> 42`. If it is not passed a promise then it will resolve immediately with the  primitive value (a way overlooked quality imo).

Seeing as action used to return nothing (or more recently actions which could be done here but there would be no point). The only issue I see with it is what @selfup mentioned which is..

> The problem with that is that Promises are not very performant even if they resolve immediately

I have no solid evidence but I have read that resolving a primitive value is evaluated as `If !IsPromise(x) return x;`, would be interesting to see because this is nice..

```javascript
function (data) {
  for (i = 0; i < hooks.onAction.length; i++) {
    hooks.onAction[i](name, data)
  }
  return Promise.resolve(app.actions[key](model, data, actions))
  .then((result) => {
    if(result) { // Check it wasn't just an effect
      for (i = 0; i < hooks.onUpdate.length; i++) {
        hooks.onUpdate[i](model, result, data)
      }
      model = merge(model, result)
      render(model, view)
    }
  })
}
```

**Comment by jbucaran** [comment]
@lukejacksonn Ah, okay, right. In that case, I don't see what prevents you from doing this already then.

**Comment by jbucaran** [comment]
@lukejacksonn It seems you were suggesting to use Promise directly inside hyperapp?

**Comment by lukejacksonn** [comment]
Yes here https://github.com/hyperapp/hyperapp/blob/master/src/app.js#L82

**Comment by jbucaran** [comment]
@lukejacksonn Even if promises were super fast, which they are not, the problem with this is forcing users to polyfill `Promise`.

**Comment by lukejacksonn** [comment]
I wasn't aware that was a limitation of the project. Most developers are very accustomed to importing [babel-polyfill](https://babeljs.io/docs/usage/polyfill) these days especially if using a bundler, it would mean having to serve an IE friendly version on CDN though perhaps.

Performance is a real issue but I would be intrigued to see how it compares. I [don't think there is going to be much in it]((https://jsperf.com/promise-vs-callback)) given the simplicity of `Promise.resolve` and how much browsers are trying to optimize it.

Let's get this PR merged and I can play around 👍

**Comment by jbucaran** [comment]
@liadbiz I've added a few comments to explain what's going on in certain places. If there are other places you think need further explanation let me know.

**Comment by jbucaran** [comment]
@lukejacksonn Thanks, for looking into perf.

See this: https://github.com/lukeed/polkadot/blob/master/benchmarks/readme.md by @lukeed.

> I wasn't aware that was a limitation of the project. Most developers are very accustomed to importing babel-polyfill these days especially if using a bundler, it would mean having to serve an IE friendly version on CDN though perhaps.

The project has a determined goal of not straying too far away from the vicinity of 1kb. If we introduce `Promise` we break this promise! 

Well, not true, if you are using an ultra modern browser, you wouldn't need a polyfill, so technically we would be 1kb, but right now, we can support IE10 and still be 1kb, which is nice.

The entire code base is a mix of mostly ES3 and some ES5 (forEach, Object.keys) and I've been really trying to keep it that way. 

Not everyone is using promises. Whatever their reasons are, if we use Promise out of the box that's a reason for a Promise-denier to not use hyperapp.

**Pull request merged**